### PR TITLE
[DDW-582] Implement wallet UTxO statistics screen

### PR DESCRIPTION
**Pull request by daniloprates**
This PR implements the wallet UTxO statistics screen.

## Todo:

- [x] Wallet settings navigation
- [x] Retrieve data from the API
- [x] UTxO statistics screen
- [x] Unit and Acceptance tests
- [x] Storybook story
- [x] [English copy and Japanese translation](https://docs.google.com/document/d/1rC4ldhoFmuxy1uK5GxmiQp7so4PWOMALALPxPhoHfDA/edit)
- [Slack proofread/translation thread](https://input-output-rnd.slack.com/archives/CBECZ2K47/p1554232895001300)
- [Slack thread for futher explanations](https://input-output-rnd.slack.com/archives/C7PJ29FME/p1554450893007300)
- [Slack thread for menu discussion](https://input-output-rnd.slack.com/archives/G7ZA8AFGD/p1555925200000700)

## Wallet has UTXOs

image

## Wallet does not have UTXOs

Screen Shot 2019-05-03 at 10 01 01

---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1554394089052500)
- [x] Test

---

### Basics

- [x] PR is updated to the most recent version of target branch (and there are no conflicts)
- [x] PR has good description that summarizes all changes and shows some screenshots or animated GIFs of important UI changes
- [x] CHANGELOG entry has been added and is linked to the correct PR on GitHub
- [x] Automated tests: All acceptance tests are passing (`yarn run test`)
- [x] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn run dev`)
- [x] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn run package` / CI builds)
- [x] There are no *flow* errors or warnings (`yarn run flow:test`)
- [x] There are no *lint* errors or warnings (`yarn run lint`)
- [x] Text changes are proofread and approved (Jane Wild)
- [x] There are no missing translations (running `yarn run manage:translations` produces no changes)
- [x] UI changes look good in all themes (Alexander Rukin)
- [x] Storybook works and no stories are broken (`yarn run storybook`)
- [x] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [x] Important parts of the code are properly documented and commented
- [x] Code is properly typed with flow
- [x] React components are split-up enough to avoid unnecessary re-rendering
- [x] Any code that only works in Electron is neatly separated from components

### Testing
- [x] New feature / change is covered by acceptance tests
- [x] All existing acceptance tests are still up-to-date
- [x] New feature / change is covered by Daedalus Testing scenario
- [x] All existing Daedalus Testing scenarios are still up-to-date

### After Review:
- [ ] Merge PR
- [ ] Delete source branch
- [ ] Move ticket to `done` on the Youtrack board

### CODE REVIEW

**Comment by daniloprates** [comment]
@nikolaglumac how often should this data be fetched? Upon wallet loading, UTxOs tab loading, any interval?

**Comment by nikolaglumac** [comment]
@daniloprates on load of this screen only. Each time user opens this screen.

**Comment by nikolaglumac** [comment]
@darko-mijic @a-rukin @daniloprates what if we would simply group the 4 largest columns on the right side of the chart and show them always?

**Comment by a-rukin** [comment]
> what if we would simply group the 4 largest columns on the right side of the chart and show them always?

but when some of those columns is not empty, we group the rest right-sided?

**Comment by nikolaglumac** [comment]
@a-rukin yes - I would always show those for as a single column (bar) regardless of their state (e.g. if they are empty or not).
cc @darko-mijic

**Comment by a-rukin** [comment]
Makes sense 
It would be quite a rare situation when those 4 last columns filled, even more rare when more than one.
Only rich guys will see that we don't show precise info there, but it seems, that they don't need to restructure that part.

**Comment by daniloprates** [comment]
> @darko-mijic @a-rukin @daniloprates what if we would simply group the 4 largest columns on the right side of the chart and show them always?

@nikolaglumac @a-rukin what would it be the label for this column? `10M+`?

**Comment by daniloprates** [comment]
@a-rukin @nikolaglumac it's not possible to change the bar color when hovering its whole area. Basically, we can set a color the whole area behind the bar, but it's not related to the bar itself:

problem

I have thought of a few solutions. Please, let me know how should I proceed. 

sol1
sol2
sol3

**Comment by a-rukin** [comment]
@daniloprates can you attach live screenshot, please?

**Comment by daniloprates** [comment]
> @daniloprates can you attach live screenshot, please?

✅

**Comment by a-rukin** [comment]
image
Question regarding this graph max value
Is it possible to have padding above the tallest column?

**Comment by daniloprates** [comment]
> image
> 
> Question regarding this graph max value.
> Is it possible to have padding above the tallest column?

@a-rukin is 10px fine?

image

**Comment by a-rukin** [comment]
@daniloprates yes, 10px seems to be nice

**Comment by a-rukin** [comment]
image
this UTxO label is better to be right-aligned with those numbers

**Comment by a-rukin** [comment]
also it's may be worth playing with those labels (and numbers) opacity
be good to have a quick screenshare session

**Comment by daniloprates** [comment]
> image
> 
> this UTxO label is better to be right-aligned with those numbers

This is a kind of [Recharts nightmare](https://github.com/recharts/recharts/issues/184). I'll give it a try and we can talk tomorrow.

**Comment by daniloprates** [comment]
> image
> 
> this UTxO label is better to be right-aligned with those numbers

@a-rukin managed to align it, but couldn't go further up due to the chart boundaries. I'll adjust it tomorrow. Also changed the opacity from 0.6 to 0.8.

Let's talk tomorrow 👋

image

**Comment by daniloprates** [comment]
> @daniloprates please make sure UTXO is always written as `UTXO` and not `UTxO` as per instructions from Amy Reeve 🙏
> cc @darko-mijic

✅

File: features/step_definitions/wallets-settings-utxos-steps.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    const [renderedWalletAmount] = await getVisibleTextsForSelector(
+      this.client,
+      selectors.walletAmount
+    );
+    const [renderedWalletUtxosAmount] = await getVisibleTextsForSelector(
+      this.client,
+      selectors.walletUtxosAmount
+    );
+
+    // TODO: Get the real active wallet amount
```
Comment (nikolaglumac) [inline features/step_definitions/wallets-settings-utxos-steps.js]: @daniloprates what about this TODO?

**Comment by daniloprates** [comment]
> @daniloprates Could you please fix the DarkTheme for the UTXO page? Thank you.
> image (20)

@a-rukin how should I deal with this? Either add a light background or make the chart elements light? Could you provide the colors for the chosen option? Thanks

File: features/step_definitions/wallets-settings-utxos-steps.js (hunk 1 above)
Comment (daniloprates) [reply features/step_definitions/wallets-settings-utxos-steps.js]: I'm having issues with this part. For some reason, in the test, `daedalus.stores.wallets.active` will return a BigNumber without its regular methods, but with Objects instead:

```shell
walletLovelaceAmount.toFormat: {}
typeof walletLovelaceAmount.toFormat: object
```

image

Comment (nikolaglumac) [reply features/step_definitions/wallets-settings-utxos-steps.js]: Interesting... We might need @DominikGuzei's help on this one 🙏

Comment (DominikGuzei) [reply features/step_definitions/wallets-settings-utxos-steps.js]: Hm … no idea why that happens, would have to debug this

**Comment by daniloprates** [comment]
@a-rukin I've simply replaced the colors by the one existing in the themes, but it looks odd. Do you have the exact colors for them?

Screen Shot 2019-04-08 at 15 30 23
Screen Shot 2019-04-08 at 15 31 00

File: features/step_definitions/wallets-settings-utxos-steps.js (hunk 1 above)
Comment (nikolaglumac) [reply features/step_definitions/wallets-settings-utxos-steps.js]: @daniloprates did you figure this one out?

File: source/renderer/app/config/timingConfig.js [hunk 2]
```diff
@@ -17,4 +17,5 @@ export const NETWORK_STATUS_POLL_INTERVAL = 2000; // 2 seconds | unit: milliseco
 export const NTP_IGNORE_CHECKS_GRACE_PERIOD = isTest ? 500 : 35 * 1000; // 35 seconds | unit: milliseconds
 export const BLOCK_CONSOLIDATION_IPC_REQUEST_INTERVAL = 10 * 1000; // 10 seconds | unit: milliseconds
 export const BLOCK_CONSOLIDATION_API_REQUEST_INTERVAL = 30 * 1000; // 30 seconds | unit: milliseconds
+export const WALLET_UTXO_API_REQUEST_INTERVAL = 10 * 1000; // 10 seconds | unit: milliseconds
```
Comment (nikolaglumac) [inline source/renderer/app/config/timingConfig.js]: @daniloprates our `WALLET_REFRESH_INTERVAL` interval is set to 5 seconds. Perhaps it is then better to use the same value here? Thanks for updating it.

**Comment by a-rukin** [comment]
@nikolaglumac pls take a look here - http://zpl.io/agQeOXM
If ok, then @daniloprates can take this in work

**Comment by nikolaglumac** [comment]
@a-rukin that looks perfect! @daniloprates please take over the colors from the link @a-rukin posted 🙏

**Comment by nikolaglumac** [comment]
PS. @daniloprates I have fixed the issue with hovering (hover area didn't match the chart column) and also added some styling improvements (positioning and centering): https://github.com/input-output-hk/daedalus/pull/1353/commits/d554272b1a17e32aa86e93aa13dd5638453bfd28

**Comment by nikolaglumac** [comment]
@daniloprates one other strange thing I found is that in this PR the colors of the Block consolidation screen on Cardano and Dark theme have gone wild 😱 

Screen Shot 2019-04-09 at 10 29 33
Screen Shot 2019-04-09 at 10 29 27

**Comment by nikolaglumac** [comment]
@daniloprates can you please merge latest develop and fix the conflicts? 🙏

File: features/step_definitions/wallets-settings-utxos-steps.js (hunk 1 above)
Comment (nikolaglumac) [reply features/step_definitions/wallets-settings-utxos-steps.js]: @daniloprates ping!

**Comment by a-rukin** [comment]
> @daniloprates the UTXO stats screen now looks and works perfectly!
> 
> Unfortunately, we have an issue with the "General" setting screen now as it simply doesn't fit in the smallest window:
> issues
> 
> @a-rukin can you please get in touch with @daniloprates and figure this one out? We need a smart solution for this layout...

re baseline - good catch! definitely should be fixed

"Transaction assurance security level"... what about TAS level with "question tooltip" ? 

For the password "change/create" can be moved on top (as in first input)

**Comment by a-rukin** [comment]
Also how graph looks in such a compressed view?

**Comment by nikolaglumac** [comment]
@a-rukin please handle this with @daniloprates - I have a lot on my plate today. I am fine with whatever you guys decide to do 🙏

**Comment by DmitriiGaico** [comment]
@daniloprates @nikolaglumac Can we adjust the name to the same display way ? Currently it looks different depending the theme you are using. 
Black
DarkBlue
Blue

**Comment by nikolaglumac** [comment]
@DmitriiGaico @daniloprates we should always show the text in a single line.

**Comment by DmitriiGaico** [comment]
> Also how graph looks in such a compressed view?

I would say the graph with skinny columns looks odd.
Compressed2
Compressed

**Comment by nikolaglumac** [comment]
@DmitriiGaico I think the chart looks good enough. However, we could try and expand the chart on the left and right to fill in the available space (seems we have enough horizontal margins to do so)...
cc @daniloprates @a-rukin

**Comment by daniloprates** [comment]
@a-rukin @nikolaglumac  my suggestion for the "Transaction assurance security level" issue:

For now, just have it as "**Security level (?)**" _(with a proper question mark icon)_ and, on hovering, expand the text to "**Transaction assurance security level**". 

But then I think we should re-think the fields and dropdown on this screen. The renaming wallet process is quite obscure for the user. I'll make a quick document with suggestions.

**Comment by nikolaglumac** [comment]
@daniloprates I think we need @darko-mijic's opinion on this: 

> For now, just have it as "Security level (?)" (with a proper question mark icon) and, on hovering, expand the text to "Transaction assurance security level".

**Comment by nikolaglumac** [comment]
@daniloprates the positioning improvements are great!
We just have a problem with Japanese text on the menu item:
issue

File: source/renderer/app/i18n/locales/en-US.json [hunk 3]
```diff
@@ -50,7 +50,7 @@
   "global.errors.invalidMnemonic": "Invalid phrase entered, please check.",
   "global.errors.invalidRepeatPassword": "Doesn't match",
   "global.errors.invalidSpendingPassword": "Invalid password",
-  "global.errors.invalidWalletName": "Wallet name requires at least 3 and at most 40 letters.",
+  "global.errors.invalidWalletName": "Wallet name requires at least 3<br />and at most 40 letters.",
```
[the remark this answers is not in the document]
Comment (daniloprates) [reply source/renderer/app/i18n/locales/en-US.json]: @MarcusHurney maybe you don't have the latest changes? We have moved the error into a tooltip, as it was hard to fit all information in that line

image

**Comment by daniloprates** [comment]
> @daniloprates @nikolaglumac
> When hovering over each column on the UTXO chart, I think it makes sense to display a value range indicating how much ada is in the UTXOs like we are currently have it, but on the last column the range shown is redundant information. I'm thinking it **should** say `1 UTXOs containing 10K or more ADA` **instead of** `1 UTXOs containing between 10K and 10K+ ADA`. What do y'all think?
> 
> Screen Shot 2019-04-25 at 8 12 35 PM

@MarcusHurney completely agree. I'll change it to your suggestion and wait for @nikolaglumac to give the final word.

File: source/renderer/app/i18n/locales/en-US.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (daniloprates) [reply source/renderer/app/i18n/locales/en-US.json]: (also, the error message renders OK on Mac, but breaks on Win & Linux 🤦🏻‍♂️)

**Comment by daniloprates** [comment]
@MarcusHurney [here](https://github.com/input-output-hk/daedalus/pull/1353/commits/c14150a4178f891fe9f131c495bc135ea32a77d7) we go:

10k

**Comment by daniloprates** [comment]
@MarcusHurney Please let me know if you can see the correct `!` tooltip icon and if there's anything else you could find. I'd be very happy to have this PR approved before the weekend :)

**Comment by MarcusHurney** [comment]
@daniloprates 
I'm not sure why the error message wouldn't show up in a tooltip for me considering I had all the changes for the UTXO screen. I'm going to check it all out again now by building a clean bundle, but beyond the error tooltip not showing up and the change you've already made to the last UTXO column, I didn't find anything else.

**Comment by MarcusHurney** [comment]
@daniloprates 
I've figured out what was going on with the tooltip not displaying the wallet name input error messages. There are a handful of places in the wallet which use the error messages you've adjusted in `en-US.json` (including the other languages as well of course). The screen shot I showed you previously where `<br />` is showing up as plain text on the label above the wallet name input is a different component from the one you added tooltip support to. 

**- Components that are rendering new error messages incorrectly -**

Create Wallet: `source/renderer/app/components/wallet/WalletCreateDialog.js`

create_wallet_name_error

Restore Paper Wallet: `source/renderer/app/components/wallet/WalletRestoreDialog.js`

restore_wallet_name_error

Restore Wallet (same component as Restore Paper Wallet but different tab): `source/renderer/app/components/wallet/WalletRestoreDialog.js`

restore_wallet_name_error

**Comment by MarcusHurney** [comment]
@daniloprates 
I've pushed a couple commits. The first one just fixes the `margin-left` on the svg icons that serve as anchors for the tooltip messages on the wallet name and security level input fields. The next commit is me adjusting the message on the last UTXO column again. I didn't realize that the value of the variable being rendered in that string contains the `+` mark. Since that is the case, there is no need to say `10K+ ADA or more` because the `+` mark literally stands for `or more` haha. It is still redundant in that sense. So I changed it to this:

Screen Shot 2019-04-27 at 4 38 15 AM

**Comment by MarcusHurney** [comment]
@daniloprates 
Ok, last one from me! I found an edge case bug in the Setting/General section of an existing wallet. If you begin editing the wallet name until it becomes invalid, let's say less than 2 characters, and then click on the create/change button for the wallet password below, it opens up the Change/Create Password dialogue. The issue is that the Wallet Name field is still focused and has an error. You aren't able to click into the input fields on the Change/Create Password Dialogue and if you try to type it will continue putting your input into the wallet name field which is hidden behind the dialogue box out of sight.

What should happen is this:
If the user clicks into the wallet name input field and enters an invalid value, then clicks onto the create/change password button, the same function that the `cancel` button runs for the wallet name input field should also run at this time. It should abort the current edit and return the value of the input field back to the last known valid value. It's simply a matter of calling the same function that is bound to the cancel button to handle this edge case.

I've attached a video showing this bug below.

[GeneralTabBug.zip](https://github.com/input-output-hk/daedalus/files/3123843/GeneralTabBug.zip)

**Comment by DmitriiGaico** [comment]
> @daniloprates
> Ok, last one from me! I found an edge case bug in the Setting/General section of an existing wallet. If you begin editing the wallet name until it becomes invalid, let's say less than 2 characters, and then click on the create/change button for the wallet password below, it opens up the Change/Create Password dialogue. The issue is that the Wallet Name field is still focused and has an error. You aren't able to click into the input fields on the Change/Create Password Dialogue and if you try to type it will continue putting your input into the wallet name field which is hidden behind the dialogue box out of sight.
> 
> What should happen is this:
> If the user clicks into the wallet name input field and enters an invalid value, then clicks onto the create/change password button, the same function that the `cancel` button runs for the wallet name input field should also run at this time. It should abort the current edit and return the value of the input field back to the last known valid value. It's simply a matter of calling the same function that is bound to the cancel button to handle this edge case.
> 
> I've attached a video showing this bug below.
> 
> [GeneralTabBug.zip](https://github.com/input-output-hk/daedalus/files/3123843/GeneralTabBug.zip)

Hello @MarcusHurney . Yeah, i already reported this really edge case scenario earlier. I'm not sure but @daniloprates created corresponding ticket for the next Sprint.

**Comment by daniloprates** [comment]
> Hello @MarcusHurney . Yeah, i already reported this really edge case scenario earlier. I'm not sure but i believe @daniloprates created corresponding ticket for the next Sprint.

Hey @DmitriiGaico, yes, card created:

image

**Comment by MarcusHurney** [comment]
@daniloprates Looks good to me! Great job sir 🥇

**Pull request merged**